bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/AbstractBigtableTable.java

```diff
@@ -298,13 +300,30 @@ public ResultScanner getScanner(final Scan scan) throws IOException {
     LOG.trace("getScanner(Scan)");
     Span span = TRACER.spanBuilder("BigtableTable.scan").startSpan();
     try (Scope scope = TRACER.withSpan(span)) {
-      clientWrapper.setCaching(scan.getCaching());
-      final ResultScanner scanner = clientWrapper.readRows(hbaseAdapter.adapt(scan));
-      if (hasWhileMatchFilter(scan.getFilter())) {
-        return Adapters.BIGTABLE_WHILE_MATCH_RESULT_RESULT_SCAN_ADAPTER.adapt(scanner, span);
+      if (scan.getCaching() == -1) {
+        final ResultScanner scanner = clientWrapper.readRows(hbaseAdapter.adapt(scan));
+        if (hasWhileMatchFilter(scan.getFilter())) {
+          return Adapters.BIGTABLE_WHILE_MATCH_RESULT_RESULT_SCAN_ADAPTER.adapt(scanner, span);
+        }
+        // TODO: need to end the span when stream ends
+        return scanner;
+      } else {
+        final Query request = hbaseAdapter.adapt(scan);
+        final RequestContext requestContext =
```

Comment:
request should already have the row limit, can we do QueryPaginator paginator = request.createPaginator(scan.getCaching) instead?

Reply:
in case request.getRowsLimit is set (like 20 in the test), and caching not set, we will end up with a request for PAGE_SIZE (scan.getCaching() is 0\-1, and is not allowed in the paginator), thus failing the test (and effectively asking for 100 rather than 20, against the user desire)

Reply:
Yeah, but QueryPagniator will only return 20 in this case. https://github.com/googleapis/java-bigtable/blob/main/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/Query.java#L388-L392

Reply:
Done